### PR TITLE
feat(hooks): Add hook decorator

### LINKED ISSUE #1483: [Feature] @hook decorator for simplified hook definitions
## Summary

Add a `@hook` decorator to enable simple, function-based hook definitions - mirroring the ergonomics of the existing `@tool` decorator. This would make hooks as easy to define and share via PyPI packages as tools are today.

## Motivation

**Current State (Class-based):**
```python
from strands.hooks import HookProvider, HookRegistry, BeforeToolCallEvent

class MyHooks(HookProvider):
    def register_hooks(self, registry: HookRegistry) -> None:
        registry.add_callback(BeforeToolCallEvent, self.on_tool_call)
    
    def on_tool_call(self, event: BeforeToolCallEvent) -> None:
        print(f"Tool: {event.tool_use}")

agent = Agent(hooks=[MyHooks()])
```

**Proposed (Decorator-based):**
```python
from strands import Agent, hook
from strands.hooks import BeforeToolCallEvent

@hook
def log_tool_calls(event: BeforeToolCallEvent) -> None:
    """Log all tool calls before execution."""
    print(f"Tool: {event.tool_use}")

agent = Agent(hooks=[log_tool_calls])
```

### Core Decorator

```python
@hook
def my_hook(event: BeforeToolCallEvent) -> None:
    """Hook description extracted from docstring."""
    pass

# Or with explicit event type
@hook(event=BeforeToolCallEvent)
def my_hook(event) -> None:
    pass
```

### Event Type Detection

1. **Type hints** (preferred): Extract from function signature
2. **Explicit parameter**: `@hook(event=BeforeToolCallEvent)`
3. **Validation**: Ensure event type inherits from `HookEvent`

### DecoratedFunctionHook Class

```python
class DecoratedFunctionHook(HookProvider):
    """Wraps a decorated function as a HookProvider."""
    
    def __init__(self, func, event_type, metadata):
        self._func = func
        self._event_type = event_type
        self._metadata = metadata
    
    def register_hooks(self, registry: HookRegistry) -> None:
        registry.add_callback(self._event_type, self._func)
    
    def __call__(self, event):
        """Allow direct invocation for testing."""
        return self._func(event)
```

### Multi-Event Hooks

```python
@hook(events=[BeforeToolCallEvent, AfterToolCallEvent])
def audit_tools(event: BeforeToolCallEvent | AfterToolCallEvent) -> None:
    """Audit both before and after tool execution."""
    pass
```

## Package Distribution

Enable sharing hooks via PyPI (like tools):

```python
# strands_observability/hooks.py
from strands import hook
from strands.hooks import BeforeInvocationEvent, AfterInvocationEvent

@hook
def trace_requests(event: BeforeInvocationEvent) -> None:
    """Start OpenTelemetry span for agent invocation."""
    pass

@hook  
def end_trace(event: AfterInvocationEvent) -> None:
    """End OpenTelemetry span."""
    pass
```

**Usage:**
```python
from strands_observability.hooks import trace_requests, end_trace

agent = Agent(hooks=[trace_requests, end_trace])
```

## API Surface

```python
# New exports from strands
from strands import hook

# New exports from strands.hooks
from strands.hooks import DecoratedFunctionHook, HookMetadata
```

## Compatibility

- **Backward compatible**: Existing `HookProvider` classes continue to work
- **Mixed usage**: Support both class-based and decorator-based hooks
- **Async support**: Handle both sync and async hook functions

```python
agent = Agent(hooks=[
    MyClassBasedHook(),  # Existing
    log_tool_calls,       # New decorator-based
])
```

## Implementation Checklist

- [ ] `FunctionHookMetadata` class for extracting function metadata
- [ ] `DecoratedFunctionHook` class implementing `HookProvider`
- [ ] `@hook` decorator with event type detection
- [ ] Support for async hook functions
- [ ] Multi-event registration support
- [ ] Documentation and examples
- [ ] Unit tests

## Related

- Tool decorator: `strands/tools/decorator.py`
- Hook system: `strands/hooks/`
- Similar pattern: `@tool` enables package distribution of tools

## Questions

1. Should `@hook` support context injection like `@tool(context=True)`?
2. Should we support hook ordering/priority via decorator params?
3. Should hooks be loadable from directories like tools (`load_hooks_from_directory`)?

### PR DESCRIPTION
**Pull request by mkmeral**
> updated description on 2/11 based on latest code

## Description

This PR adds a `@hook` decorator that transforms Python functions into `HookProvider` implementations with automatic event type detection from type hints.

## Motivation

Defining hooks currently requires implementing the `HookProvider` protocol with a class, which is verbose for simple use cases:

```python
# Current approach - verbose
class LoggingHooks(HookProvider):
    def register_hooks(self, registry: HookRegistry) -> None:
        registry.add_callback(BeforeToolCallEvent, self.on_tool_call)

    def on_tool_call(self, event: BeforeToolCallEvent) -> None:
        print(f"Tool: {event.tool_use}")

agent = Agent(hooks=[LoggingHooks()])
```

The `@hook` decorator provides a simpler function-based approach that reduces boilerplate while maintaining full compatibility with the existing hooks system.

Resolves: #1483

## Public API Changes

New `@hook` decorator exported from `strands` and `strands.hooks`:

```python
from strands import Agent, hook
from strands.hooks import BeforeToolCallEvent

@hook
def log_tool_calls(event: BeforeToolCallEvent) -> None:
    print(f"Tool: {event.tool_use}")

agent = Agent(hooks=[log_tool_calls])
```

Event types are automatically detected from the function's type hints. The decorator supports the following patterns:

```python
# Single event via type hint
@hook
def my_hook(event: BeforeToolCallEvent) -> None: ...

# Multiple events via Union type hint
@hook
def my_hook(event: BeforeToolCallEvent | AfterToolCallEvent) -> None: ...

# Async hooks
@hook
async def my_hook(event: BeforeToolCallEvent) -> None: ...

# Class methods
class MyHooks:
    @hook
    def my_hook(self, event: BeforeToolCallEvent) -> None: ...

```

## Related Issues

Fixes #1483

## Documentation PR

No documentation changes required.

## Type of Change

New feature

## Testing

- Added comprehensive unit tests (53 test cases)
- Tests cover: basic usage, union types, async, class methods, descriptor protocol, error handling, edge cases
- [x] I ran `hatch run prepare`

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [x] I have updated the documentation accordingly
- [x] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/strands/hooks/decorator.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+def hook(__func: F) -> DecoratedFunctionHook[Any]: ...
+
+
+@overload
+def hook(
+    *,
+    event: type[BaseHookEvent] | None = None,
+    events: Sequence[type[BaseHookEvent]] | None = None,
+) -> Callable[[F], DecoratedFunctionHook[Any]]: ...
+
```
Comment (afarntrog) [inline src/strands/hooks/decorator.py]: I really like this decorator idea!

However, I think there are too many ways to use it. From the description I see:
```python
# Type hint detection
@hook
def my_hook(event: BeforeToolCallEvent) -> None: ...

# Explicit event type
@hook(event=BeforeToolCallEvent)
def my_hook(event) -> None: ...

# Multiple events via parameter
@hook(events=[BeforeToolCallEvent, AfterToolCallEvent])
def my_hook(event) -> None: ...

# Multiple events via Union type
@hook
def my_hook(event: BeforeToolCallEvent | AfterToolCallEvent) -> None: ...

# Async hooks
@hook
async def my_hook(event: BeforeToolCallEvent) -> None: ...

# Class methods
class MyHooks:
    @hook
    def my_hook(self, event: BeforeToolCallEvent) -> None: ...
```




If we look a the [Zen of Python](https://peps.python.org/pep-0020/) we see:
> There should be one-- and preferably only one --obvious way to do it.
which I very much agree with.

I suggest we trim it down to passing in the events just with the `events` parameter like so:
```python

# Explicit event type
@hook(events=[BeforeToolCallEvent])
def my_hook(event) -> None: ...

# Multiple events via parameter
@hook(events=[BeforeToolCallEvent, AfterToolCallEvent])
def my_hook(event) -> None: ...
```

Comment (mkmeral) [reply src/strands/hooks/decorator.py]: > I suggest we trim it down to passing in the events just with the events parameter like so

I get the point, but I dont like it, because if I am using type hints (for autocomplete, mypy, etc); why do I need to duplicate the events I handle, i.e. write it in both method definition and hook methods?

**Review by cagataycali — APPROVED** [review]
## Review Summary

This is an excellent implementation of the `@hook` decorator. The design is clean, follows Python best practices, and provides a significantly improved developer experience.

### What's Good ✅

1. **Elegant API Design**: The decorator supports multiple usage patterns seamlessly:
   - Simple type hint inference: `@hook` + typed parameter
   - Explicit single event: `@hook(event=EventType)`
   - Multiple events via list: `@hook(events=[...])`
   - Union type inference: `event: A | B`

2. **Proper Type Handling**: Handles both `typing.Union` and Python 3.10+ `types.UnionType` for `A | B` syntax.

3. **Descriptor Protocol**: Correctly implements `__get__` for proper method binding in class contexts.

4. **Comprehensive Tests**: 53 test cases covering edge cases including async hooks, class methods, error scenarios, and agent integration.

5. **Preserves Function Metadata**: Uses `functools.update_wrapper` to maintain introspection.

### Minor Observations

The implementation handles the "no type hint" error case well - it guides users to use `@hook(event=EventType)` when type hints aren't available.

### Alignment with SDK Tenets

- **"Obvious path is happy path"** ✅ - `@hook` with type hints is intuitive
- **"Simple at any scale"** ✅ - Works for single functions or complex class-based hooks
- **"Accessible to humans and agents"** ✅ - Clean API, good docstrings

This reduces boilerplate significantly while maintaining full compatibility with the existing hooks system. Great work @mkmeral!

---
🤖 *AI agent response from the Strands team. [Strands Agents](https://github.com/strands-agents). Feedback welcome!*

File: tests/strands/hooks/test_decorator.py [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+        assert original_function.__name__ == "original_function"
+        assert original_function.__doc__ == "Original docstring."
+
+
+class TestAsyncHooks:
+    """Tests for async hook support."""
+
+    def test_async_hook_direct_invocation(self):
+        """Test async hook direct invocation."""
+        import asyncio
```
Comment (poshinchen) [inline tests/strands/hooks/test_decorator.py]: nit: static import

File: src/strands/hooks/decorator.py (hunk 1 above)
Comment (JackYPCOnline) [reply src/strands/hooks/decorator.py]: I think we should limit it a bit...

Comment (mkmeral) [reply src/strands/hooks/decorator.py]: I've removed the params in decorator. It works only with typehints right now 👍👍

**Comment by mkmeral** [comment]
Closing in favor of https://github.com/strands-agents/sdk-python/pull/1740
